src/server/game/Entities/Creature/Creature.cpp

```diff
@@ -77,68 +77,64 @@ VendorItem const* VendorItemData::FindItemCostPair(uint32 item_id, uint32 extend
     return nullptr;
 }
 
-uint32 CreatureTemplate::GetRandomValidModelId() const
+CreatureModel const* CreatureTemplate::GetModelByIdx(uint32 idx) const
 {
-    uint8 c = 0;
-    uint32 modelIDs[4];
-
-    if (Modelid1) modelIDs[c++] = Modelid1;
-    if (Modelid2) modelIDs[c++] = Modelid2;
-    if (Modelid3) modelIDs[c++] = Modelid3;
-    if (Modelid4) modelIDs[c++] = Modelid4;
-
-    return ((c>0) ? modelIDs[urand(0, c-1)] : 0);
+    return Models.size() <= idx ? &Models[idx]: nullptr;
 }
 
-uint32 CreatureTemplate::GetFirstValidModelId() const
+CreatureModel const* CreatureTemplate::GetRandomValidModel() const
 {
-    if (Modelid1) return Modelid1;
-    if (Modelid2) return Modelid2;
-    if (Modelid3) return Modelid3;
-    if (Modelid4) return Modelid4;
-    return 0;
-}
+    if (!Models.size())
+        return nullptr;
 
-uint32 CreatureTemplate::GetFirstInvisibleModel() const
-{
-    CreatureModelInfo const* modelInfo = sObjectMgr->GetCreatureModelInfo(Modelid1);
-    if (modelInfo && modelInfo->is_trigger)
-        return Modelid1;
+    // If only one element, ignore the Probability (even if 0)
+    if (Models.size() == 1)
+        return &(*Models.begin());
 
-    modelInfo = sObjectMgr->GetCreatureModelInfo(Modelid2);
-    if (modelInfo && modelInfo->is_trigger)
-        return Modelid2;
+    auto selectedItr = Trinity::Containers::SelectRandomWeightedContainerElement(Models, [](CreatureModel const& model)
+    {
+        return model.Probability;
```

Comment:
It's not really Probability, it's more like VisiblePlayer. All triggers now have a displayid hidden (1) and a display for GM (0)

Reply:
I can confirm it's Probability

INSERT INTO `creature_template_model` (`entry`, `Idx`, `CreatureDisplayID`, `DisplayScale`, `Probability`, `VerifiedBuild`) VALUES
(26125, 3, 24995, 1, 0.25, 27377), -- 26125
(26125, 2, 24992, 1, 0.25, 27377), -- 26125
(26125, 1, 24993, 1, 0.25, 27377), -- 26125
(26125, 0, 24994, 1, 0.25, 27377), -- 26125

(131477, 3, 81836, 1, 0.5, 27377), -- 131477
(131477, 2, 81853, 1, 0.5, 27377), -- 131477
(131477, 1, 82334, 1, 1, 27377), -- 131477
(131477, 0, 82335, 1, 1, 27377), -- 131477

Reply:
Ah yeah, misunderstanding. So when it's 0, it's a display for GM for sure

Reply:
If 0 and there is multiple models yes

If 0 and there is only one model, it's the default display